src/loops/surrounding.lean

```diff
@@ -17,15 +20,115 @@ local notation `I` := Icc (0 : ℝ) 1
 def loop.surrounds (γ : loop F) (x : F) : Prop := 
   ∃ t w : fin (d + 1) → ℝ, surrounding_pts x (γ ∘ t) w
 
+lemma loop.surrounds_iff_range_subset_range (γ : loop F) (x : F) : 
+  γ.surrounds x ↔ ∃ (p : fin (d + 1) → F) (w : fin (d + 1) → ℝ), 
+  surrounding_pts x p w ∧ range p ⊆ range γ :=
+begin
 
+  split,
+  { exact λ ⟨t, w, h⟩, ⟨(γ ∘ t), w, h, range_comp_subset_range _ _⟩ },
+  { rintros ⟨p, w, h₀, h₁⟩,
+    rw range_subset_iff at h₁,
+    choose t ht using h₁,
+    have hpt : γ ∘ t = p := funext ht,
+    exact ⟨t, w, hpt.symm ▸ h₀⟩ }
+end
+
+set_option profiler true
 lemma surrounding_loop_of_convex_hull {f b : F} {O : set F} (O_op : is_open O) (O_conn : is_connected O) 
   (hsf : f ∈ convex_hull O) (hb : b ∈ O) : 
   ∃ γ : ℝ → loop F, continuous_on ↿γ (set.prod I univ) ∧ 
-                    ∀ t, γ t 0 = b ∧
-                    ∀ s, γ 0 s = b ∧
-                    ∀ (t ∈ I) s, γ t s ∈ O ∧
+                    (∀ t, γ t 0 = b) ∧
+                    (∀ s, γ 0 s = b) ∧
+                    (∀ (t ∈ I) s, γ t s ∈ O) ∧
                     (γ 1).surrounds f :=
-sorry
+begin
```

Comment:
The length of this proof says the API is not there yet. We probably need many more preparatory lemmas.

Reply:
Although it's clearly too long, I'm not sure about which parts I should factor out. Most intermediary results seem too specific to be a separate lemma

Reply:
I think you need more `def` both `Prop` valued and data. You should probably define 'round_trip` (either as a propery of `loop` or a structure extending `loop`, then a function which takes a round trip loop and turn it into a family of loops interpolating between a constant loop and the given round trip. Then you can state lemmas about these constructions and you will have separated completely the round trip trick from building a path through a family of points.

Reply:
Would just defining `round_trip` as a specific loop built from a path (say `loop.of_path (γ.trans γ.symm)` for some `γ : path x y`) ? I don't really see the point of making it a general predicate/structure here, because I don't need to prove things about all round trip loops, but maybe I'm missing something

Reply:
Since this is an internal definition, it depends what you end up using in the proof. You need to try, keeping in mind the goal is to retract these loops to constant loops.